src/esp/physics/bullet/BulletDebugManager.cpp

```diff
@@ -0,0 +1,186 @@
+  std::stringstream result;
+
+  result << static_cast<const void*>(colObj) << " " << name << ", "
+         << activationStateNames[colObj->getActivationState()];
+
+  const auto* broadphaseHandle = colObj->getBroadphaseHandle();
+  if (broadphaseHandle) {
+    result << ", group "
+           << std::to_string(broadphaseHandle->m_collisionFilterGroup) +
+                  ", mask "
+           << std::to_string(broadphaseHandle->m_collisionFilterMask);
+  } else {
+    // this probably means the object hasn't been added to the bullet world
+    result << ", no broadphase handle";
+  }
+  return result.str();
+}
+
+std::string BulletDebugManager::getStepCollisionSummary(
+    btMultiBodyDynamicsWorld* bWorld) {
+  std::stringstream s;
+  auto func = [&](const btCollisionObject* colObj0,
+                  const btCollisionObject* colObj1,
+                  const btPersistentManifold* manifold) {
+    if (manifold->getNumContacts() > 0) {
+      s << "[" << getDebugStringForCollisionObject(colObj0) << "] vs ["
+        << getDebugStringForCollisionObject(colObj1) << "], "
+        << manifold->getNumContacts() << " points" << std::endl;
+    }
+  };
+
+  processActiveManifolds(bWorld, func);
+
+  const bool isEmpty = s.rdbuf()->in_avail() == 0;
+  if (isEmpty) {
+    s << "(no active collision manifolds)" << std::endl;
+  }
+
+  return s.str();
+}
+
+std::string BulletDebugManager::getCollisionFilteringSummary(bool doVerbose) {
+  std::stringstream s2;
+  for (const auto& pair0 : collisionObjectToDebugName_) {
+    const btCollisionObject* colObj0 = pair0.first;
+    s2 << getDebugStringForCollisionObject(colObj0) << std::endl;
+  }
+
+  if (doVerbose) {
+    for (const auto& pair0 : collisionObjectToDebugName_) {
+      std::stringstream s;
+      const btCollisionObject* colObj0 = pair0.first;
+      s << getDebugStringForCollisionObject(colObj0)
+        << " may collide with:" << std::endl;
+
+      bool foundAny = false;
+      for (const auto& pair1 : collisionObjectToDebugName_) {
+        const btCollisionObject* colObj1 = pair1.first;
+        const std::string& name1 = pair1.second;
+        if (colObj1 == colObj0) {
+          continue;
+        }
+
+        const auto* broadphaseHandle0 = colObj0->getBroadphaseHandle();
+        const auto* broadphaseHandle1 = colObj1->getBroadphaseHandle();
+
+        // logic copied from
+        // btHashedOverlappingPairCache::needsBroadphaseCollision
+        bool collides = (broadphaseHandle0->m_collisionFilterGroup &
+                         broadphaseHandle1->m_collisionFilterMask) != 0;
+        collides = collides && (broadphaseHandle1->m_collisionFilterGroup &
+                                broadphaseHandle0->m_collisionFilterMask);
+
+        if (collides) {
+          s << "  " << name1 << std::endl;
+          foundAny = true;
+        }
+      }
+
+      if (!foundAny) {
+        s << "  (none)" << std::endl;
+      }
+
+      LOG(INFO) << s.str();
```

Comment:
This logging is a horrible hack I did as a workaround at one point (specifically, I had trouble printing very long strings to stdout). This function shouldn't directly log; it should return all results in the string.